### PR TITLE
feat: LT-2146 - Contract types

### PR DESCRIPTION
**Pull request by cstobart**

### CODE REVIEW

File: applications/serializers/open_application.py [hunk 1]
```diff
@@ -94,8 +95,12 @@ def get_destinations(self, application):
                 )
                 .order_by("-contains_flags", "highest_flag_priority", "name")
             )
+        qs = CountryOnApplication.objects.filter(application=application)
+
+        coa_list = [qs.get(country_id=c.id) for c in countries]
+
+        serializer = CountryOnApplicationViewSerializer(coa_list, many=True, context={"flags": True})
```
Comment (efahy) [inline applications/serializers/open_application.py:102]: Can we keep the name `active_flags_only` in the context please?  It is more indicative of the purpose of it

File: applications/tests/tests_contract_types_on_countries.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+from applications.models import CountryOnApplication
+from cases.libraries.get_flags import get_ordered_flags
+from flags.tests.factories import FlagFactory
+from lite_content.lite_api import strings
+from static.countries.models import Country
+from test_helpers.clients import DataTestClient
+
+
+class ContractTypeOnCountryTests(DataTestClient):
+    @tag("2146")
```
Comment (efahy) [inline applications/tests/tests_contract_types_on_countries.py:15]: Don't forget to remove these tags

File: applications/tests/tests_contract_types_on_countries.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+        data = {"countries": ["GB"], "contract_types": ["other_contract_type"], "other_contract_type_text": ""}
+
+        url = reverse("applications:contract_types", kwargs={"pk": application.id})
+
+        response = self.client.put(url, data, **self.exporter_headers)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    @tag("2146")
+    def test_set_other_contract_type_on_country_on_application_success(self):
```
Comment (efahy) [inline applications/tests/tests_contract_types_on_countries.py:58]: Perhaps one more test that the API will reject requests where other isn't a selected contract type, but they have sent other text?

File: applications/models.py [hunk 4]
```diff
@@ -318,6 +320,11 @@ class CountryOnApplication(models.Model):
     id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
     application = models.ForeignKey(OpenApplication, related_name="application_countries", on_delete=models.CASCADE)
     country = models.ForeignKey(Country, related_name="countries_on_application", on_delete=models.CASCADE)
+    contract_types = SeparatedValuesField(
+        max_length=350, choices=ContractType.choices, blank=True, null=True, default=None
```
[the remark this answers is not in the document]
Comment (cstobart) [reply applications/models.py]: done

File: applications/serializers/open_application.py (hunk 1 above)
Comment (cstobart) [reply applications/serializers/open_application.py:102]: done

File: applications/tests/tests_contract_types_on_countries.py (hunk 2 above)
Comment (cstobart) [reply applications/tests/tests_contract_types_on_countries.py:15]: I won't

File: applications/tests/tests_contract_types_on_countries.py (hunk 3 above)
Comment (cstobart) [reply applications/tests/tests_contract_types_on_countries.py:58]: ok

File: applications/views/countries.py [hunk 5]
```diff
@@ -30,15 +33,26 @@ def get(self, request, application):
...
+        print("Time to serialize ", len(countries), "countries: ", time_4 - time_3)
+        time_omega = datetime.now()
+        print("Time to get countries heavy: ", time_omega - time_alpha)
         return JsonResponse(data={"countries": countries_data}, status=status.HTTP_200_OK)
 
     @transaction.atomic
     @allowed_application_types([CaseTypeSubTypeEnum.OPEN])
     @authorised_users(ExporterUser)
     def post(self, request, application):
+        print("\nPosting countries")
```
Comment (jfaracik) [inline applications/views/countries.py:54]: Bye bye prints

File: test_helpers/clients.py [hunk 6]
```diff
@@ -176,17 +176,18 @@ def setUp(self):
                 ]
             ]
         )
-
-        if settings.TIME_TESTS:
+        if True:
```
Comment (jfaracik) [inline test_helpers/clients.py:179]: Stuff here!

**Pull request merged**